### ISSUE TITLE
Autogenerated code for fromJson in all models doesn't match AppSync's query results when there are nested models in the query

### ISSUE THREAD

**Issue by epomeranz** [aws-amplify/amplify-flutter#816]
**Describe the bug**
Amplify automatically generated the code for all the models, including the fromJson functions. Additionally amplify also generates the basic resolvers for basic queries. Unfortunately, the amplify queries do not work with the fromJson functions when there are nested tables...
i.e. if I have this query:

```graphql
query MyQuery {
  listConversations {
    items {
      id
      type
      messages {
        items {
          message
          createdAt
        }
      }
      participants {
        items {
          usersID
        }
      }
    }
  }
}
```

Conversation has a list of messages and a list of participants... In the query result messages and participants are nested under the "items" key, and not under the root keys for "participants" and "messages". This is OK because you are including additional information under those root keys. But the automatically generated fromJson functions for all models expects that nested queries be on the root and not under an "items" key...

Here is the fromJson autogenerated code:
```dart
messages = json['messages'] is List
        ? (json['messages'] as List)
          .where((e) => e?['serializedData'] != null)
          .map((e) => Message.fromJson(new Map<String, dynamic>.from(e['serializedData'])))
          .toList()
        : null,
      _participants = json['participants'] is List
        ? (json['participants'] as List)
          .where((e) => e?['serializedData'] != null)
          .map((e) => Participant.fromJson(new Map<String, dynamic>.from(e['serializedData'])))
          .toList()
        : null,
```

This should be:
```dart
messages = json['messages'][items] is List
        ? (json['messages'][items] as List)
          .where((e) => e?['serializedData'] != null)
          .map((e) => Message.fromJson(new Map<String, dynamic>.from(e['serializedData'])))
          .toList()
        : null,
      _participants = json['participants'][items] is List
        ? (json['participants'][items] as List)
          .where((e) => e?['serializedData'] != null)
          .map((e) => Participant.fromJson(new Map<String, dynamic>.from(e['serializedData'])))
          .toList()
        : null,
```
**To Reproduce**
1. Create at leas 2 entities. One nested under the other in a (0 to many relationship)
2. Autogenerate the flutter code for the models
3. Run a query that reads the parent and the many children (0 to many relationship)
4. The fromJson code will ignore the nested models (children) because they are under an "items" key.

**Expected behavior**
The fromJson code should expect the same structure that the queries provide. They are both part of amplify so a read and a fromJson should work seamlessly and not have different structures.

**Screenshots**

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[X] Android
[X] iOS

**Comment by Jordan-Nelson** [comment]
Hi @epomeranz - Thanks for taking the time to open the issue. Do you mind sharing your `schema.graphql` file and the snippets of dart/flutter code that you are using to perform the save/query operations?

**Comment by epomeranz** [comment]
Hi Jordan,

Unfortunately I can't share the graphQL schema for the app but I don't think the bug is limited to my schema. The sample schema will have the same inconsistency between the fromJson autogenerated code and the actual graphQL entities. The actual schema nests lists inside an "items" element. The fromJson expects the list to be in the root element and not nested inside an "items" element...

Is there any other way I can help without sharing the code?

**Comment by epomeranz** [comment]
I found another similar bug but in the "opposite direction". When an entity (i.e. "Message") as a sub-entity (i.e. "Owner") that is NOT a list (so not owners but a message has one single owner) then the fromJson expect the owner to be nested under a "serializedData" entity. So the autogenerated fromJson would look like this:
_owner = json['owner']?['serializedData'] != null... (You can see that the fromJson function is expecting the data to be under json['owner']?['serializedData'] and not just under json['owner'].

However, Amplify.API.query(request: request); returns the data only under "owner" with no 'serializedData'...

As a result the fromJson doesn't read any sub-entity...

**Comment by rahul-readwhere** [comment]
Hey, 
I am just curious you mentioned ," Additionally amplify also generates the basic resolvers for basic queries". I have been stuck on this for a long time to write a query in flutter to list filtered items in graphql. I know it is a bit off topic but do you mind suggesting something

**Comment by epomeranz** [comment]
Hi Rahul,

Here is a tip: Go to the AWS appSync console for your project and select the "queries" screen. Here you should be able to create queries using a UI. You can then copy/paste the queries to flutter and run them using the Amplify.API
That will get you 90% there. Unfortunately, the documentation for flutter amplify queries is almost non-existent so you'll have to do a lot of try and "error" to get the final query working.
I wrote try and "error" with quotes because there is not "error" only try. Since the amplify SDK for flutter doesn't throw errors if you do something wrong. It just doesn't work and sometimes stops the execution of the current thread. So you have no feedback if you are doing something wrong. It's actually the worst SDK I've used in my 20+ years of development :(
If you can the best approaches I found are to try to use the REST API with amplify and avoid the SDK. Or even better, if you can, avoid using AWS with flutter. It's currently not mature enough for production usage. Try Supabase, firebase/GCP, or anything else...

**Comment by OLTApp** [comment]
I am having the same issues as well. I using the latest amplify cli .4.0 and flutter 2.10.2, dart 2.16.1.  

- As stated above it has ["serializedData"] which does not exists from the response from AppSync API
- For list the generated code is missing ["items"]

I manually verified the response by running the query using AppSync query UI page.

**Comment by ragingsquirrel3** [comment]
While this issue is not technically resolved, there is a change in 0.4.2 that should help. While model helpers like `ModelQueries.list()` pre-generate a list request with no children, you can include some metadata on a custom list request (even one that's deeply nested) and the result will be correctly typed with children so you don't need to touch the `fromJson()` at all. Here is an example with the classic blog > post > comment schema:

```dart
String graphQLDocument = '''query GetBlogsCustomDecoder {
  listBlogs {
    items {
      id
      name
      createdAt
      posts {
        items {
          id
          title
          rating
          comments {
            items {
              id
              content
            }
          }
        }
      }
    }
  }
}''';
final request = GraphQLRequest<PaginatedResult<Blog>>(
    document: graphQLDocument,
    modelType: const PaginatedModelType(Blog.classType),
    decodePath: 'listBlogs');
final response = await Amplify.API.query(request: request).response;
final blogWithNestedData = response.data?.items.first;
```

By including the fields `modelType` and `decodePath`, the plugin knows how to convert it correctly, the same as it would for a pre-generated request.

I'm going to leave this issue open as the fromJson still cannot be used directly for a similar purpose as @epomeranz stated. The plugin code is doing a transform internally to handle this. The `fromJson` in the codegen is also used to convert the serialized data from the datastore plugin on the platform side, so its format is intended to match that. To change it, we would have to also change some things in datastore to match it.

**Comment by OLTApp** [comment]
Thanks ! this helps a lot to keep the original model, now I can use both DataStore.Subscription and API. Once we have the nested model support for hasMany, hasOne etc, data store will be in a solid position for lot of use cases.

**Comment by MarlonJD** [comment]
There's still an issue continues. I cannot list manyToMany's second field. I can get on AppSync console, but it's coming null on ModelQueries.list. even when I'm trying to create with Model mutations.create. it's second field of ManyToMany field becoming null, and i modify the variables and force to create this id. So this is a huge problem...

**Comment by exelnait** [comment]
Still an issue in cli 10.5.2

**Comment by anchals333** [comment]
Still exist in cli 10.7.3!

**Comment by MarlonJD** [comment]
> Still exist in cli 10.7.3!

Are you using stable or next version ?

**Comment by anchals333** [comment]
> Are you using stable or next version ?

I'm using stable

**Comment by MarlonJD** [comment]
Can you share, your flutter version, amplify versions, sample schema and code? I'll try to reproduce error

**Comment by samuelrbrt** [comment]
This is still an issue when including a custom query with nested data.

`  Article.fromJson(Map<String, dynamic> json)  
    : id = json['id'],
      _title = json['title'],
      _author = json['author']?['serializedData'] != null
        ? User.fromJson(new Map<String, dynamic>.from(json['author']['serializedData']))
        : null,
      _banner = json['banner']?['serializedData'] != null
        ? Media.fromJson(new Map<String, dynamic>.from(json['banner']['serializedData']))
        : null,
      _contents = json['contents'] is List
        ? (json['contents'] as List)
          .where((e) => e?['serializedData'] != null)
          .map((e) => ArticleContent.fromJson(new Map<String, dynamic>.from(e['serializedData'])))
          .toList()
        : null,
      _categoryId = json['categoryId'],
      _tags = json['tags']?.cast<String>(),
      _readTime = json['readTime'],
      _commentsCount = (json['commentsCount'] as num?)?.toInt(),
      _isActive = json['isActive'],
      _createdAt = json['createdAt'] != null ? amplify_core.TemporalDateTime.fromString(json['createdAt']) : null,
      _updatedAt = json['updatedAt'] != null ? amplify_core.TemporalDateTime.fromString(json['updatedAt']) : null,
      _articleAuthorId = json['articleAuthorId'],
      _articleBannerKey = json['articleBannerKey'];`

Why is the generated type tries to read ['serializedData'] ? which is not there in the response at all! What a pain to use Flutter SDK to integrate API to apps!

**Comment by 0xAPPA** [comment]
any updates on this ? fromJson() still doesn't work with nested fields

**Comment by luiscib3r** [comment]
I had to deal with this problem recently. I put the use case in context and how I solved it.

First I had to make a custom query to AppSync:

```dart
    const queryName = 'userEvaluationsByDocumentIdAndOwnerId';

    final document = '''
    query ListUserEvaluations {
      ${documentsIds.asMap().map(
              (index, documentId) => MapEntry('q$index', '''
      $queryName(
        documentId: "$documentId",
        ownerId: {eq: "$ownerId"}
      ) {
        items {
          id
          ownerId
          choice {
            description
            points
          }
          evaluationMethodId
        }
      }
      '''),
            ).entries.map((e) => "${e.key}: ${e.value}").join('\n')}
    }
    ''';

    final request = GraphQLRequest<String>(
      document: document,
    );
```

What I do is perform the `userEvaluationsByDocumentIdAndOwnerId` query but for all the ids found in documentsIds and I assign the alias `q0`, `q1`, `q2`, etc. to each query. Finally the request is created and I make the request to obtain the response.

```dart
final response = await Amplify.API.query(request: request).response;
```

The response comes as a string in the response.data field so I convert it to Map<String, dynamic>

But in this case `choice` is a nested object. And it happens that the code generated for UserEvaluation.fromJson is trying to do this:

``dart
_choice = json['choice']?['serializedData'] != null
             ? EvaluationChoice.fromJson(
                 new Map<String, dynamic>.from(json['choice']['serializedData']))
             :null,
```
Which is incorrect.

```
final data = json.decode(response.data ?? '{}') as Map<String, dynamic>;
```

Now each `q0`, `q1`, etc is a `PaginatedModelType<UserEvaluation>`. Normally it can be parsed as follows.

```dart
final evaluations = PaginatedModelType(UserEvaluation.classType)
          .fromJson(data['q0'])
          .items
          .where((e) => e != null)
          .cast<UserEvaluation>()
          .toList();
```

But in this case `choice` is a nested object. And it happens that the code generated for UserEvaluation.fromJson is trying to do this:

```dart
_choice = json['choice']?['serializedData'] != null
             ? EvaluationChoice.fromJson(
                 new Map<String, dynamic>.from(json['choice']['serializedData']))
             :null,
```

Which is incorrect.

If we check the static `classType` attribute of any generated model, in this case UserEvaluation. We can see something similar to this:

```dart
class _UserEvaluationModelType extends amplify_core.ModelType<UserEvaluation> {
   const _UserEvaluationModelType();

   @override
   UserEvaluation fromJson(Map<String, dynamic> jsonData) {
     return UserEvaluation.fromJson(jsonData);
   }

   @override
   String modelName() {
     return 'UserEvaluation';
   }
}
```

My solution then was to override the `fromJson` method by creating a new `classType` to use in custom queries.

```dart
class UserEvaluationParser extends ModelType<UserEvaluation> {
  @override
  UserEvaluation fromJson(Map<String, dynamic> jsonData) {
    return UserEvaluation(
      id: jsonData['id'] as String,
      ownerId: jsonData['ownerId'] as String,
      choice: EvaluationChoice.fromJson(
        jsonData['choice'] as Map<String, dynamic>,
      ),
      evaluationMethodId: jsonData['evaluationMethodId'] as String,
    );
  }
}
```

Then change the json parser code to:

```dart
final evaluations = PaginatedModelType(UserEvaluationParser())
          .fromJson(data['q0'])
          .items
          .where((e) => e != null)
          .cast<UserEvaluation>()
          .toList();
```

In summary, the solution is to create a new class that extends `ModelType` and use it to override the `fromJson` method that is generated by default. You can then use `PaginatedModelType(YourCustomParser()).fromJson` to parse the json data.

**Comment by anchals333** [comment]
+1, I'm facing the same issue & handled it  by creating my custom models as per response for now.

**Comment by IvanWarpSpeedTechDev** [comment]
I am still facing this issue. It is now 2024. There should be a fix for this as this is a standard use case for graphql. Come on guys. please.

**Comment by Jordan-Nelson** [comment]
Hello folks - Apologies for the delay in response. We are working on this issue and expect to have a version of the library available soon that resolved this. We will post here when it is available.

**Comment by Equartey** [comment]
Hi Y'all! This has been addressed in Amplify Flutter V2. Please follow our [upgrade guide](https://docs.amplify.aws/gen1/flutter/start/project-setup/upgrade-guide/) to take full advantage 💪 

Depending on your use case, Amplify CLI `12.12.2` also includes new a `.fromJson()` branch that enables decoding of AppSync responses directly. 

If you have any questions or concerns, don't hesitate to reach out as usual.
